modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java

```diff
@@ -327,9 +334,14 @@ protected void onNodeLeft(UUID evtNodeId) {
     protected void enqueue(Collection<?> col) {
         assert Thread.holdsLock(this);
 
-        queue.add((Collection<R>)col);
-
-        cnt.addAndGet(col.size());
+        if(limit <= 0 || total.addAndGet(col.size()) <= limit) {
```

Comment:
Is it possible to rewrite this with no adding 'total' field, but using CAS-loop or cnt.getAndUpdate() ?

Reply:
'total' counter was added because 'cnt' used here as queue size indicator and is incremented on inserts and decremented when item is consumed. So we cannot rely on it as on total counter.